### ISSUE TITLE
How to properly restart a somehow stopped job??

### ISSUE THREAD

**Issue by iz3man** [cncjs/cncjs#795]
This is not a cncjs specific "issue", and may somehow apply to other cnc programs as well, but I use cncjs, and look for a proper way to restart a broken job.

Eg. The grbl nc file created has some issues and triggers and end switch which makes the machine go into ALARM. Then you need to RESET and work coordinates are set to 0/0/0 again with the spindle at some completetly different location.
Or a mill breaks during operation and you need to restart it. or ....

How do I "save" the work coordinates? So I do a homing, which hopefully is precise enough to give me consistant 0/0/0 position of the mill. And then I move the spindle to the position I need and zero out in cncjs webfrontend and start the job.

I know could write down the spindle machine's coordinates and move to that position once the machine got and issue.

But isn't there a better way to do it?
I'm pretty sure this is just a complete noob's question and easy to be answered?!

Thanks!

**Comment by MitchBradley** [comment]
I have never had much success in restarting interrupted jobs. The interruption itself is very likely to ruin the part.The one exception is when the job is an array of identical parts, so the ruined one can be discarded and the rest cut from the remaining stock. For that purpose, I have versions of my gcode program that do all 9 parts, only the last 8, only the last 7, etc.  An alternative is to rerun the entire job “air cutting” the ones that are already complete, but that requires precise rezeroing so you do not mar good parts that are still attached to the stockOn Jan 7, 2023, at 5:08 AM, iz3man ***@***.***> wrote:﻿
This is not a cncjs specific "issue", and may somehow apply to other cnc programs as well, but I use cncjs, and look for a proper way to restart a broken job.
Eg. The grbl nc file created has some issues and triggers and end switch which makes the machine go into ALARM. Then you need to RESET and work coordinates are set to 0/0/0 again with the spindle at some completetly different location.
Or a mill breaks during operation and you need to restart it. or ....
How do I "save" the work coordinates? So I do a homing, which hopefully is precise enough to give me consistant 0/0/0 position of the mill. And then I move the spindle to the position I need and zero out in cncjs webfrontend and start the job.
I know could write down the spindle machine's coordinates and move to that position once the machine got and issue.
But isn't there a better way to do it?
I'm pretty sure this is just a complete noob's question and easy to be answered?!
Thanks!

**Comment by iz3man** [comment]
Thanks Mitch. My main issue is 're-zeroing' as you call it. 

I just now had a broken end mill at the end of a 2 part's job. The machine returned to homing after the job was done. I replaced the mill and generated a new nc file which included only the cuts not done. That went very well as the machine still had homeing and work pieces coordinates intact.

The question is how to remove the "alarm state" w/o changing the work coordinates. Is that even possible?

**Comment by MitchBradley** [comment]
If you are using Grbl and the alarm entry type requires a hard reset via Ctrl-X (that is the character that is sent to Grbl; I don't remember what button you have to press on CNCjs to send that and I do not have time to look it up), then Grbl is going to do whatever it does with a Ctrl-X reset, which might change the work coordinates.  I do not think that it changes them, but I am not sure.

**Comment by MitchBradley** [comment]
When you said RESET, you did not specify how you did that.  Was it a power cycle of the controller, or did you press a physical reset button, or did you send Ctrl-X via some UI button?  If you want a good answer, you have to ask a precise question.

**Comment by iz3man** [comment]
It seems RESET can be done in different ways, and it SEEMS outcome is different?!
I found the best way to use is RESET button in the WFE. 

![image](https://user-images.githubusercontent.com/11553674/211162080-cc7ffb88-896d-4278-b163-084cc1b469eb.png)


But still this somehow alters the working coordinates, especially Z axis it seems.
Resetting directly on the machine is the worst as this doesn't seem to be communicated back to cncjs.

**Comment by cheton** [comment]
It is recommended to run the HOMING operation before starting each new job to ensure a consistent starting position. 

Additionally, I remember that Snapmaker 2 has built-in support for resuming interrupted jobs due to power failure. 

I think @parachvte can help answer if the resuming part is easy to do on the software side.
